gazelle_cabal/dependency_resolution.go

```diff
@@ -223,85 +214,74 @@ func getPackageLabel(
 	pkgName string,
 	from label.Label,
 ) label.Label {
-	var resolvers []func(*resolve.RuleIndex, string, string, label.Label)(label.Label, error)
-	resolvers = append(resolvers,
-		resolveAgainstPrivateLibraries,
-		resolveAgainstPublicLibraries,
-		resolveAgainstRepository,
-	)
-
-	for _, resolver := range resolvers {
-		lbl, err := resolver(ix, packageRepo, pkgName, from)
-		if (err == nil) {
-			return lbl
+	for _, searchScope := range obtainSearchScopes(pkgName, from) {
+		if labelFound, err := searchInLibraries(ix, packageRepo, pkgName, from, searchScope); err == nil {
+			return labelFound
 		}
 	}
 
-	log.Fatalf("Could not resolve dependency '%s' from '%s'", pkgName, from)
-	return label.Label{}
+	return grabFromRepository(packageRepo, pkgName, from)
 }
 
-func resolveAgainstPrivateLibraries(
-	ix *resolve.RuleIndex,
-	packageRepo string,
-	pkgName string,
-	from label.Label,
-) (label.Label, error) {
-	parentPackage, depName := splitDependencyString(pkgName)
-
-	if len(parentPackage) > 0 && (parentPackage != depName) {
-		return label.Label{}, fmt.Errorf("It looks like %s is not an internal, private library", pkgName)
-	}
-
-	// Search for the rule of an internal library using the prefix "internal_library:" for the key
-	spec := resolve.ImportSpec{gazelleCabalName, "private_library:" + depName}
-	res := ix.FindRulesByImport(spec, gazelleCabalName)
-
-	// Search for the label of an internal library in the current package
-	for _, r := range res {
-		if r.IsSelfImport(from) {
-			log.Fatalf("Dependency cycle detected in the following component: %s", from)
-		}
-
-		// indeed internal library can be refereced form the package of its definition
-		if (r.Label.Repo == from.Repo && r.Label.Pkg == from.Pkg) {
-			return rel(r.Label, from), nil
+// Dependency string may contain 'experimental' colon syntax e.g package-b:sublib
+func obtainSearchScopes(depName string, from label.Label) []string {
+	splitted := strings.Split(depName, ":")
+	publicPrefix := "public_library"
+	privatePrefix := "private_library"
+	format := "%s:%s:%s"
+	if len(splitted) > 1 {
+		cabalPackageId := splitted[0]
+		libraryId      := splitted[1]
+		return []string{
+			// there is prefix which can reference the public or internal library
+	        fmt.Sprintf(format, publicPrefix, cabalPackageId, libraryId),
+			// TODO: potentially throw error if from.Pkg != cabalPackageId?
+			fmt.Sprintf(format, privatePrefix, from.Pkg, libraryId),
```

Comment:
If the dependency string contains a packageId (say "tasty:tasty"), it doesn't seem correct to discard it and look for an internal library instead ("package-a:tasty"). Is this code still correct?

Reply:
Ahh I think I know what you mean now... I'm not 100% familiar with cabal but I see that it is allowed define the import for some `respository` dependency in such way (with colon syntax as you did: `tasty:tasty`). In such case we should take tasty from repository, right? I'll have to do something with it. So you're right - this code doesn't cover all the cases.

But just to explain what it does right now. Assuming:
package-a.cabal
```
library sublib
 ...

library
  build-depends:
    package-b:sublib
```
package-b.cabal
```
library sublib:
  visibility:public
```
Order matters here, therefore on the very beginning I'm searching for *public* libraries instead of private ones. If the order would be opposite, than `package-b:sublib` in package-a.cabal would be resolved by us to package-a:sublib.